File: Solutions/JavaScript/49. Group Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const mp = new Map();
  let res = [];

  for (let i = 0; i < strs.length; i++) {
    let x = strs[i].split("").sort().join("");
    if (mp.has(x)) {
      mp.get(x).push(strs[i]);
    } else mp.set(x, [strs[i]]);
  }

  mp.forEach((val)=>{
    res.push(val)
  })

  return res;
};

// Solution 2: Brute Force
// var isAnagram = function(s, t) {
//   if (s.length !== t.length) {
//       return false;
//   }
  
//   let freq = new Array(26).fill(0);
//   for (let i = 0; i < s.length; i++) {
//       freq[s.charCodeAt(i) - 'a'.charCodeAt(0)]++;
//       freq[t.charCodeAt(i) - 'a'.charCodeAt(0)]--;
//   }
  
//   for (let i = 0; i < freq.length; i++) {
//       if (freq[i] !== 0) {
//           return false;
//       }
//   }
  
//   return true;
// };
// var groupAnagrams = function(strs) {
//   let arr = [];
//   let done = {};
//   for(let i = 0; i < strs.length; i++){
//       let temp = [];
//       if (!done[strs[i]]){
//           for(let j=0;j<strs.length;j++){
//               if(isAnagram(strs[i],strs[j])){
//                   temp.push(strs[j]);
//                   done[strs[j]] = true;
//               }
//           }
//       }
//       if(temp.length)
//       arr.push(temp);
//   }
//   return arr;
// };

File: Solutions/Kotlin/49-group-anagrams.kt
/*
 * Runtime: 298 ms, faster than 94.13% of Kotlin online submissions for Group Anagrams.
 * Memory Usage: 43.7 MB, less than 39.88% of Kotlin online submissions for Group Anagrams.
*/

class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        
        var i = 0
        var map = hashMapOf<String,ArrayList<String>>()
        var result = arrayListOf<List<String>>()
        
        // 1-This while loop goes through each string in the strs array
        while(i<strs.size){
            var actual =  strs[i]
            
            // 2-sorts its characters alphabetically
            var arr = actual.toCharArray()
            arr.sort()
            
            // 3-stores it as a new string called sortedText.
            var sortedText = String(arr)
            
            /* 4- checks if this sortedText key is already present in the map HashMap.
            If it is, it retrieves the ArrayList of strings that corresponds to that key,
            adds the current actual string to the list, and puts the list back into the map.*/
            var actualList =  map.getOrDefault(sortedText, arrayListOf<String>())
            
            /* 5- If the key is not present in the map, 
            it creates a new empty ArrayList of strings and puts it in the map with the current sortedText key.*/
            actualList.add(actual)
            map.put(sortedText,actualList)
            //println(arr)
            i++
        }
        
        
        /*6-This forEach loop goes through each key-value pair in the map HashMap,
        and adds the corresponding ArrayList of strings to the result ArrayList of Lists of Strings.*/
        map.forEach{key, item->
        
            result.add(item)
        }
        
        return result
    }
}
